daffodil-lib/src/main/scala/org/apache/daffodil/cookers/EntityReplacer.scala

```diff
@@ -568,7 +568,9 @@ trait NoCharClassEntitiesMixin {
    * Override if set of prohibited class entities is different
    */
   protected def noCharClassEntities(raw: String, context: ThrowsSDE): Unit = {
-    context.schemaDefinitionUnless(!raw.contains("%NL;"), "Property dfdl:%s cannot contain %%NL;", propName)
+    if(!raw.contains("%%")){
+      context.schemaDefinitionUnless(!raw.contains("%NL;"), "Property dfdl:%s cannot contain %%NL;", propName)
+    }
```

Comment:
This is actually a really tricky and subtle situtation. Because of this, I think this isn't quite right for the check to fix DAFFODIL-774.

As an example, textStandardZeroRep property is not allowed to contain the NL and ES character classes. So something like this is not allowed:
```xml
<xs:element ... dfdl:textStandardZeroRep="0_%NL;" />
```
In this example we have a zero rep of the string ``0``, followed by an underscore, followed by a newline. I believe this works as intented right now and creates an SDE because this contains an unallowed ``%NL;``

However, someone could theoretically do something like this (note the escaping `%%`` before NL):
```xml
<xs:element ... dfdl:textStandardZeroRep="0_%%NL;" />
```
So the zero rep in this case is the string ``0``, followed by an underscore, followed by the literal string ``%NL;`` instead of a newline. We would currently consider that an SDE because the raw string contains ``%NL;``, even though that's not really a ``%NL;`` because the ``%`` is escaped via a double ``%``. Note that it is very unlikely for a user to ever want a literal ``%NL;`` as a property value, so one might even argue making this check better as DAFFODIL-744 describes isn't worth it. Though it is an interesting excercise, and would make Daffodil more correct if we did add it.

However, what your check would do is make it so that if the raw property value contains ``%%`` then we don't even bother checking for ``%NL;`` at all. So for example, say we had something like this:
```xml
<xs:element ... dfdl:textStandardZeroRep="0%%_%NL;" />
```
So in this case, the user specified that the zero rep is the string ``0``, followed by a single perect sign (%% escape means single perect), followed by an underscore, followed by a newline. This shouldn't be allowed because the zero rep contains ``%NL;``. But because it contains %%, your new check would skip the ``%NL;`` check and it won't error as it should. Again, this is a contrived example, but that's what this edge case is talking about.

To properly fixed DAFFODIL-744, I don't think it's enough to just do simple string searches on the raw value like we do now. How we detect %NL; and other disallowed character classes must change, because it's too difficult to take into account all the different escape combinations. I think instead of looking at the raw value, we need to inspect the cooked value (or what we do with the cooked value). I'm not too familar with that code, but I think cooking the strings is supposed to deal with things like escaped percents, character classes, etc. and so the cooked result should have that escaping done and provide information to actually know if the property actually contains an unescaped %NL; or not. I'm not positive about this though.

If it's not clear where to get starting on making that change, it might be helpful to start a discussion on the dev list and see if there are any other ideas on how to best fix this, or maybe we'll decide it's not even worth the effort. It is a very rare edge case.

Whatever the solution, we also must make sure it works for ES, WSP, WSP+ and other character classes. Although the bug only lists NL as a problem, this same escaping issue exists for those properties.

Also, if you do continue working on this, I would reccommend starting with writing some TDML tests (maybe using the examples I've mentioned above as a starting point) to show we have the incorrect behavior, and what the correct behavior should be for some more complex cases with lots of escaping going on. It we determine the fix isn't worth the effort, we would at least have tests showing the expected behavior and how Daffodil doesn't do the right thing in case we ever do decide to fix it.

Reply:
I think you want to modify NoCharClassEntitiesMixin to make the checks do more than just search for %NL; or similar, because that does not exclude "%%NL;" which is allowed. 

The correct predicate test really has to use a regex to find an "unescaped %". The regex for this is quite tricky. It has to prevent an odd number of consecutive % characters. E.g., "%%%NL;" is %% then %NL;, so should cause an SDE. But "%%%%" is %%, %%, NL; which is allowed. A regex for "unescaped %" is the key concept, and this odd/even number of % is the core idea, and that can definitely be done with a regex.

BTW: my regex goto links are:
http://myregextester.com/
http://rick.measham.id.au/paste/explain.pl
https://www.regular-expressions.info/lookaround.html

Reply:
Ah yeah, that's much simpler than what I was thinking. A even/odd check should do it just fine.